File: template_method/post_account_controller.go
package template_method

type PostAccountController[RQ PostAccountReqBody, RS PostAccountResBody, Dto string] struct {
	*BasePostController[RQ, RS, Dto]
	dto      Dto
	response RS
}

func NewPostAccountController[RQ PostAccountReqBody, RS PostAccountResBody, Dto string]() *PostAccountController[RQ, RS, Dto] {
	postAccountController := &PostAccountController[RQ, RS, Dto]{BasePostController: &BasePostController[RQ, RS, Dto]{}}
	postAccountController.IBasePostController = postAccountController
	return postAccountController
}

func (p PostAccountController[RQ, RS, Dto]) CreateAccount(requestBody RQ) *Response {
	return p.BasePostController.Execute(requestBody)
}

func (p PostAccountController[RQ, RS, Dto]) MapToDto(requestBody RQ) Dto {
	reqBody := PostAccountReqBody(requestBody)
	return Dto(reqBody.BranchCode + reqBody.AccountNumber)
}

func (p *PostAccountController[RQ, RS, Dto]) DoService(accountNumber Dto) RS {
	return RS(*NewPostAccountResBody(string(accountNumber)))
}

func (p PostAccountController[RQ, RS, Dto]) CreateResponse(responseBody RS) *Response {
	return NewResponse(responseBody, 201)
}

File: template-method/post-account.controller.ts
import { BasePostController } from "./base-post-controller";
import { PostAccountReqBody } from "./post-account-req-body";
import { PostAccountResBody } from "./post-account-res- body";
import { Response } from "./response";

export class PostAccountController extends BasePostController<PostAccountReqBody, PostAccountResBody, string> {
    public createAccount(requestBody: PostAccountReqBody): Response<any> {
        return this.execute(requestBody);
    }

    protected mapToDto(requestBody: PostAccountReqBody): string {
        return requestBody.branchCode + requestBody.accountNumber;
    }

    protected doService(accountNumber: string): PostAccountResBody {
        return new PostAccountResBody(accountNumber);
    }

    protected createResponse(responseBody: PostAccountResBody ): Response<PostAccountResBody> {
        return new Response(responseBody, 201);
    }
}